### PR TITLE
Supporting SpringMapMessageListener

### LINKED ISSUE #710: Should add support for instrumenting the listeners with parameter as subTypes of javax.jms.Message (i.e. javax.jms.MapMessage)
JMS-Plugin

Existing JMS listeners are instrumented only if the parameter to onMessage is "javax.jms.Message". The "javax.jms.MapMessage" which is also frequently used subType isn't handled yet. We should consider supporting the suptypes of "javax.jms.Message"

SessionAwareMessageListeners from springframework are also widely used. Not having support for that can restrict the usage of jms plugin.

There are few messaging providers who have class names like "QuesSender" for example "TibjmsQueueSender.TibjmsQueueSender" from EMS which aren't supported in the current version.

### PR DESCRIPTION
**Pull request by eyalkoren**
closes elastic/apm-agent-java#710

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java [hunk 1]
```diff
@@ -199,6 +199,8 @@ public boolean asChildOf(String traceParentHeader) {
         } catch (IllegalArgumentException e) {
             logger.warn(e.getMessage());
             return false;
+        } finally {
+            onMutation();
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: I think it is a bug without that, right?

Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: Is that when re-using a TraceContext? Could you add a test case for this scenario?

File: apm-agent-plugins/apm-jms-plugin/pom.xml [hunk 2]
```diff
@@ -24,21 +27,32 @@
...
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-jms</artifactId>
+            <version>${spring-framework.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-context</artifactId>
+            <version>${spring-framework.version}</version>
```
Comment (felixbarny) [inline apm-agent-plugins/apm-jms-plugin/pom.xml]: `<scope>provided</scope>`?

File: apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java [hunk 3]
```diff
@@ -63,14 +63,15 @@ public JmsMessageListenerInstrumentation(ElasticApmTracer tracer) {
 
     @Override
     public ElementMatcher<? super TypeDescription> getTypeMatcher() {
-        return not(isInterface()).and(hasSuperType(named("javax.jms.MessageListener")));
+        return not(isInterface())
```
Comment (felixbarny) [inline apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java]: This formatting may make it a bit clearer than the unnecessary parenthesis? YMMV

```
return not(isInterface())
    .and(
        hasSuperType(named("javax.jms.MessageListener"))
            .or(hasSuperType(named("org.springframework.jms.listener.SessionAwareMessageListener")))
    );
```

File: apm-agent-plugins/apm-jms-plugin/pom.xml (hunk 2 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-jms-plugin/pom.xml]: No, test! (only relies on JMS API, the Spring class is only used by name)
Thanks!

File: apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java (hunk 3 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java]: Yes, there is an extra unnecessary pair

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: No, it's when I set the Transaction a child of remote span after it is created. The tests pass without it, but I noticed in the logs that it was wrong. Will add a test.

**Comment by eyalkoren** [comment]
@felixbarny Added to CHANGELOG, but I am not sure about adding to supported technologies.
I assume most users utilizing this Spring feature assume this is standard JMS..

**Pull request merged**